File: api/proxy.js
export default async function handler(req, res) {
  const targetHost = 'mathjoy.eu.org';         // ← your main site
  const aliasHost = req.headers.host;    // alias.com
  const targetUrl = `https://${targetHost}${req.url}`;

  // Copy request headers, override Host
  const headers = { ...req.headers, host: targetHost, origin: `https://${targetHost}` };

  // Forward the request
  const response = await fetch(targetUrl, {
    method: req.method,
    headers,
    body: ['GET','HEAD'].includes(req.method) ? undefined : req.body,
    redirect: 'manual'
  });

  // Copy response headers
  response.headers.forEach((value, key) => {
    if (!['content-length','connection'].includes(key.toLowerCase())) {
      res.setHeader(key, value);
    }
  });

  // Cache static resources
  const contentType = response.headers.get('content-type') || '';
  if (/(image|javascript|css|font|svg)/i.test(contentType)) {
    res.setHeader('Cache-Control', 'public, max-age=86400'); // 1 day
  } else {
    res.setHeader('Cache-Control', 'no-cache'); // HTML no-cache
  }

  // Rewrite HTML links from main.com → alias.com
  if (contentType.includes('text/html')) {
    let html = await response.text();
    html = html.replaceAll(targetHost, aliasHost);
    res.status(response.status).send(html);
  } else {
    res.status(response.status);
    const reader = response.body.getReader();
    const writer = res;
    const stream = new ReadableStream({
      async start(controller) {
        while(true){
          const { done, value } = await reader.read();
          if(done) break;
          controller.enqueue(value);
        }
        controller.close();
      }
    });
    return new Response(stream).body.pipeTo(writer);
  }
}

File: api/proxy.py
import httpx
from urllib.parse import urljoin
from fastapi import FastAPI, Request, Response

app = FastAPI()

TARGET_HOST = "mathjoy.eu.org"  # your Cloudflare-hosted domain

@app.api_route("/{path:path}", methods=["GET", "POST", "PUT", "DELETE", "OPTIONS", "PATCH", "HEAD"])
async def proxy(request: Request, path: str):
    # Build target URL
    target_url = urljoin(f"https://{TARGET_HOST}/", path)
    query = request.url.query
    if query:
        target_url += "?" + query

    # Prepare headers (set Host header to target)
    headers = dict(request.headers)
    headers["Host"] = TARGET_HOST

    # Send request to Cloudflare-protected site
    async with httpx.AsyncClient(follow_redirects=True) as client:
        resp = await client.request(
            method=request.method,
            url=target_url,
            headers=headers,
            content=await request.body()
        )

    # If HTML, replace domain references
    content_type = resp.headers.get("content-type", "")
    body = resp.content
    if "text/html" in content_type:
        body = body.replace(TARGET_HOST.encode(), request.url.hostname.encode())

    # Return proxied response
    return Response(
        content=body,
        status_code=resp.status_code,
        headers={k: v for k, v in resp.headers.items() if k.lower() not in ["content-encoding", "transfer-encoding"]}
    )
